examples/pytorch/graphsage/experimental/train_dist_unsupervised.py

```diff
@@ -0,0 +1,393 @@
+                h_dst = h[:block.number_of_dst_nodes()]
+                h = layer(block, (h, h_dst))
+                if l != len(self.layers) - 1:
+                    h = self.activation(h)
+                    h = self.dropout(h)
+
+                y[output_nodes] = h.cpu()
+
+            x = y
+            g.barrier()
+        return y
+
+def load_subtensor(g, input_nodes, device):
+    """
+    Copys features and labels of a set of nodes onto GPU.
+    """
+    batch_inputs = g.ndata['features'][input_nodes].to(device)
+    return batch_inputs
+
+class CrossEntropyLoss(nn.Module):
+    def forward(self, block_outputs, pos_graph, neg_graph):
+        with pos_graph.local_scope():
+            pos_graph.ndata['h'] = block_outputs
+            pos_graph.apply_edges(fn.u_dot_v('h', 'h', 'score'))
+            pos_score = pos_graph.edata['score']
+        with neg_graph.local_scope():
+            neg_graph.ndata['h'] = block_outputs
+            neg_graph.apply_edges(fn.u_dot_v('h', 'h', 'score'))
+            neg_score = neg_graph.edata['score']
+
+        score = th.cat([pos_score, neg_score])
+        label = th.cat([th.ones_like(pos_score), th.zeros_like(neg_score)]).long()
+        loss = F.binary_cross_entropy_with_logits(score, label.float())
+        return loss
+
+def generate_emb(model, g, inputs, batch_size, device):
+    """
+    Generate embeddings for each node
+    g : The entire graph.
+    inputs : The features of all the nodes.
+    batch_size : Number of nodes to compute at the same time.
+    device : The GPU device to evaluate on.
+    """
+    model.eval()
+    with th.no_grad():
+        pred = model.inference(g, inputs, batch_size, device)
+
+    return pred
+
+def compute_acc(emb, labels, train_nids, val_nids, test_nids):
+    """
+    Compute the accuracy of prediction given the labels.
+    """
+
+    emb = emb[np.arange(labels.shape[0])].cpu().numpy()
+    train_nids = train_nids.cpu().numpy()
+    val_nids = val_nids.cpu().numpy()
+    test_nids = test_nids.cpu().numpy()
+    labels = labels.cpu().numpy()
+
+    emb = (emb - emb.mean(0, keepdims=True)) / emb.std(0, keepdims=True)
+    lr = lm.LogisticRegression(multi_class='multinomial', max_iter=10000)
+    lr.fit(emb[train_nids], labels[train_nids])
+
+    pred = lr.predict(emb)
+    f1_micro_eval = skm.f1_score(labels[val_nids], pred[val_nids], average='micro')
+    f1_micro_test = skm.f1_score(labels[test_nids], pred[test_nids], average='micro')
+    f1_macro_eval = skm.f1_score(labels[val_nids], pred[val_nids], average='macro')
+    f1_macro_test = skm.f1_score(labels[test_nids], pred[test_nids], average='macro')
+    return f1_micro_eval, f1_micro_test, f1_macro_eval, f1_macro_test
+
+def run(args, device, data):
+    # Unpack data
+    train_eids, train_nids, in_feats, g, global_train_nid, global_valid_nid, global_test_nid, labels = data
+    # Create sampler
+    sampler = NeighborSampler(g, [int(fanout) for fanout in args.fan_out.split(',')], train_nids,
```

Comment:
this is weird. the negative edges are corrected by the nodes in `train_nids`? why are they corrupted by the nodes in the training set instead of all nodes in the local partition?